numpy/core/src/multiarray/calculation.c

```diff
@@ -150,6 +151,58 @@ PyArray_ArgMax(PyArrayObject *op, int axis, PyArrayObject *out)
     return NULL;
 }
 
+NPY_NO_EXPORT PyObject*
+PyArray_ArgMaxKeepdims(PyArrayObject *op, int axis, PyArrayObject* out) {
+    /*
+    * `axis_copy` retains a copy of original axis value because, `axis`
+    * itself is changed by the call in the following line. This is done
+    * for two reasons, first, `axis_copy` helps in checking whether axis
+    * was `None` or not as received from Python interface and second, axis
+    * values are consistent between `PyArray_ArgMax` and `PyArray_ArgMaxKeepdims`.
+    */
+    int axis_copy = axis;
+    if ((PyArrayObject *)PyArray_CheckAxis(op, &axis, 0) == NULL) {
+        return NULL;
+    }
+    PyArrayObject* ret = (PyArrayObject*)PyArray_ArgMax(op, axis_copy, NULL);
+    PyArray_Dims newdims;
+    newdims.len = PyArray_NDIM(op);
+    npy_intp dims[NPY_MAXDIMS];
+    newdims.ptr = dims;
+    npy_intp* currdim = PyArray_DIMS(op);
+    for( int dim = 0; dim < newdims.len; dim++ ) {
+        /*
+        * Here `axis` is converted to 0 when `axis = NPY_MAXDIMS`
+        * but `axis_copy` retains that value and hence, all values
+        * in `newdims.ptr` can be set to 1 if axis is received as `None`
+        * from Python interface.
+        */
+        if( dim == axis || axis_copy == NPY_MAXDIMS ) {
+            newdims.ptr[dim] = 1;
+        } else {
+            newdims.ptr[dim] = currdim[dim];
+        }
+    }
+    
+    // If ret is same as out, then should we reshape or raise an error?
+    ret = (PyArrayObject*)PyArray_Newshape(ret, &newdims, NPY_CORDER);
+    if (out != NULL) {
+        if ((PyArray_NDIM(out) != PyArray_NDIM(ret)) ||
+                !PyArray_CompareLists(PyArray_DIMS(out), PyArray_DIMS(ret),
+                                      PyArray_NDIM(out))) {
+            PyErr_SetString(PyExc_ValueError,
+                    "output array dimensions do not match with that of result from np.argmax.");
+            Py_XDECREF(ret);
+            return NULL;
+        }
+        PyArray_CopyInto(out, ret);
+        Py_DECREF(ret);
+        ret = out;
+        Py_INCREF(ret);
+    }
+    return (PyObject*)ret;
+}
+
```

Comment:
This approach uses to reshape to match the number of dimensions with that of the input.